notebooks/analyze_looping_across_lending_protocols.ipynb

```diff
@@ -0,0 +1,989 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import pandas as pd\n",
+    "import mysql.connector\n",
+    "from IPython.display import display  # Only needed for Jupyter Notebook\n",
+    "\n",
+    "# Connect to the MySQL database\n",
+    "conn = mysql.connector.connect(\n",
+    "    host='your_host',\n",
+    "    user='your_username',\n",
+    "    password='your_password',\n",
+    "    database='loans_db'\n",
+    ")\n",
+    "\n",
+    "# List of protocols (table names in the MySQL database)\n",
+    "protocols = [\"zklend\", \"nostra_alpha\", \"nostra_mainnet\", \"hashstack_v0\", \"hashstack_v1\"]\n",
+    "\n",
+    "for protocol in protocols:\n",
+    "    print(f\"Processing {protocol}...\")\n",
+    "    \n",
+    "    # Query the data from the MySQL database\n",
+    "    query = f\"SELECT * FROM {protocol}\"\n",
+    "    df = pd.read_sql_query(query, conn)\n",
+    "    \n",
+    "    pd.set_option('display.max_columns', None)  # Display all columns\n",
+    "    pd.set_option('display.max_colwidth', None)  # Display full column width\n",
+    "    pd.set_option('display.width', None)  # Adjust display width\n",
+    "\n",
+    "    # Display the first rows\n",
+    "    display(df.head())\n",
+    "\n",
+    "# Close the connection\n",
+    "conn.close()\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### From GCS"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import pandas as pd\n",
+    "import pyarrow.parquet as pq\n",
+    "import requests\n",
+    "import matplotlib.pyplot as plt\n",
+    "import seaborn as sns\n",
+    "from io import BytesIO\n",
+    "\n",
+    "# URLs of the loans files for all lending protocols\n",
+    "parquet_urls = {\n",
+    "    \"zklend\": \"https://storage.googleapis.com/derisk-persistent-state/zklend_data/loans.parquet\",\n",
+    "    \"nostra_alpha\": \"https://storage.googleapis.com/derisk-persistent-state/nostra_alpha_data/loans.parquet\",\n",
+    "    \"nostra_mainnet\": \"https://storage.googleapis.com/derisk-persistent-state/nostra_mainnet_data/loans.parquet\",\n",
+    "    \"hashstack_v0\": \"https://storage.googleapis.com/derisk-persistent-state/hashstack_v0_data/loans.parquet\",\n",
+    "    \"hashstack_v1\": \"https://storage.googleapis.com/derisk-persistent-state/hashstack_v1_data/loans.parquet\"\n",
+    "}\n",
+    "\n",
+    "dataframes = {}\n",
+    "\n",
+    "for protocol,url in parquet_urls.items():\n",
+    "    # Download the file\n",
+    "    response = requests.get(url)\n",
+    "    response.raise_for_status()  # Ensure the request was successful\n",
+    "\n",
+    "    # Read the Parquet file into a Pandas DataFrame\n",
+    "    with BytesIO(response.content) as f:\n",
+    "        table = pq.read_table(f)\n",
+    "        df = table.to_pandas()\n",
```

Comment:
```suggestion
    "        loan_states = table.to_pandas()\n",
```

Reply:
Done! ✅

Reply:
Change it also on other relevant places (MySQL, PostgreSQL) if applicable. This holds for my other comments as well.